### PR TITLE
Validate Bearer JWT for authentication

### PR DESCRIPTION
**Pull request by bickelj**
<details><summary>(Outdated original message)</summary>

~~As of this commit, only for authentication, but eventually should be authorization as well.~~

~~This is a progress commit demonstrating a change using the /applicants endpoint. It should work in conjuction with a Bearer token issued by keycloak using the OAuth 2.0 Device Authorization Grant.~~
</details>

Expect a JSON Web Token (JWT) in the `Authorization` header for
protected resources. Validate the JWT using the configured realm key.
This allows the standard OpenID Connect protocol to be used for
authentication to an auth server, namely Keycloak, and the resulting
token to be valid for calls to this service. For example, a user of a
browser-based UI can log in using the standard Authorization Code flow
and that UI can make calls to this back-end service using the JWT.

For audit trail purposes, it is still important to log some portion of
the JWT. A JWT, while expected to be short-lived, is still useful as
credentials in its complete form. Part of the validation of a JWT is
signature validation. To keep the audit trail while preventing replay
from a server log, redact the signature of the JWT in logs.

The OpenAPI documentation is a normal browser-based Client. This
change expects the documentation Client to have a particular name,
'pdc-openapi-docs', to be configured in the 'pdc' realm of the
authentication service.

Issue #96 Authentication

### CODE REVIEW

**Comment by bickelj** [comment]
There are some awkward things in here that will need cleanup, such as the TypeScript types coming into production/non-dev dependencies, the weird need to wrap the expressjwt function or else fail linting, and probably more. Also it does not check if a user is in a group but it validates that the token was indeed issued by the issuer configured at server launch time. I pushed in order to show progress as well as get feedback on this approach.

**Comment by bickelj** [comment]
8e0bd2e has a rebase and also an additional commit demonstrating authorization (on top of authentication) from the application perspective.

**Comment by bickelj** [comment]
It seems we agree keycloak is the right free/libre open source software auth solution. We are still working on the best way to integrate keycloak. This PR might be the way to go but there is also an alternative (or complement) https://authjs.dev/ to consider.

**Comment by bickelj** [comment]
The push of 83285ab includes a rebase on main, updates to dependencies, and removal of the `ErrorWithStatus` (to be considered separately if needed).

File: package-lock.json
Comment (bickelj) [inline package-lock.json:17]: Repeated dependency declarations here, whoops. I am surprised npm did not error on this.

I am still trying to figure out why several `@types` dependencies are ending up outside the dev dependencies area when doing `npm install`.

File: package-lock.json
Comment (bickelj) [inline package-lock.json:3374]: This could be the culprit, new non-dev dependency `express-jwt` declares dependency on `@types/jsonwebtoken` without saying that one should be dev dependency.

File: package-lock.json
Comment (bickelj) [inline package-lock.json:5244]: Likewise here: `jwks-rsa`, a dependency, does not say these are dev dependencies.

Comment (bickelj) [reply package-lock.json:5244]: Hmm, found this comment: https://github.com/auth0/node-jwks-rsa/pull/80#pullrequestreview-244880314
Which references this: https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#dependencies
I thought most libraries let downstream libraries use the typescript types optionally, though.

**Comment by bickelj** [comment]
a9b72d1 rebases on main, updates dependencies, and manually puts the `@types/*` dependencies in dev dependencies by adding the dev flag in `package-lock.json` because it sounds like auth0 is going to keep those in production dependencies.

**Comment by bickelj** [comment]
The branch at 45db64a has been rebased on main again.

File: src/auth/jwtOptions.ts
Comment (jasonaowen) [inline src/auth/jwtOptions.ts:27]: Can you say more about this? Why is this better than the default?

**Review by jasonaowen — COMMENTED** [review]
Could you please rebase this on `main` and resolve the merge conflicts, @bickelj?

**Comment by bickelj** [comment]
@jasonaowen Thank you for the review. This last push was not yet in response to your review but may have incidentally resolved a couple of your comments. I will address your comments after taking care of an update to #164.

File: src/__tests__/applicants.int.test.ts
Comment (bickelj) [reply src/__tests__/applicants.int.test.ts]: I don't see them in my most recent push, so probably not.

File: package.json
Comment (bickelj) [reply package.json:57]: Before [the suggested change above to the type of jwtOptions](https://github.com/PhilanthropyDataCommons/service/pull/124#discussion_r1128674483) there was only a `type` import from `jsonwebtoken`, right? So it would come from `@types/jsonwebtoken` and all the `@types` are only `devDependencies`. Granted, `jsonwebtoken` is a transitive dependency of one of our `dependencies` (`espress-jwt`), so it does end up in the `dependencies` in `package-lock.json`.

Now with [the suggested change above to the type of jwtOptions](https://github.com/PhilanthropyDataCommons/service/pull/124#discussion_r1128674483) there is no direct dependency on `@types/jsonwebtoken` nor `jsonwebtoken` in `src/auth/jwtOptions.ts`, but there is one import in `src/middleware/jwtTokenChecker.ts` of the `JwtPayload` type which is needed at compile time. Therefore `@types/jsonwebtoken` can stay at compile time (`devDependencies` in package.json) but we don't need a reference to plain `jsonwebtoken` at all in the `package.json` It might still arrive in `package-lock.json` in non-dev `dependencies` due to being a transitive dependency, though.

File: src/auth/jwtOptions.ts
Comment (bickelj) [reply src/auth/jwtOptions.ts:27]: I am trying to find where I got this (or why I put it there), but looking again, I do not see why it should be included. The only possible advantage we could get would be if we used our own Error type but we do not. I will see if anything bad happens when removing it.

File: src/middleware/jwtTokenChecker.ts
Comment (jasonaowen) [reply src/middleware/jwtTokenChecker.ts]: I do still prefer, please! You're probably right that we do need authz as part of the pilot, but either way I think it'll be easier if they're separated.

File: package.json
Comment (jasonaowen) [reply package.json:57]: I'll look at this more deeply later, but in general the policy is: if you `import` (not `import type`) from a dependency in non-test code, it goes in `dependencies`, even if it's already a transitive dependency; if you `import` from a dependency in test code, or `import type` from a dependency in non-test code, it goes in `devDependencies`; if you never import it, then leave it to your direct dependencies to bring it in as a transitive dependency (except in some special cases).

I think you already know that, but I mention it just in case; more likely, what's happening is that there's been a bit of thrash on this branch and we're talking about different states of the world!

**Comment by jasonaowen** [comment]
We'll also need to update the Swagger docs as part of this, since it currently knows about the API key header.

**Comment by bickelj** [comment]
160cf6f should pass lint but not tests. It is still a work in progress.

File: package.json
Comment (bickelj) [reply package.json]: It sounds like we're on the same page. I think this is resolved. If not, we can revisit.

File: README.md
Comment (jasonaowen) [inline README.md:133]: I've found that it is clearer and more sustainable to put environment variable documentation in the `.env.example` instead of the README. It's closer to what it's documenting, it's more likely to be updated correctly as environment variables change, and it simplifies the instructions in the README. I did that recently in the [data-viewer `.env.example`](https://github.com/PhilanthropyDataCommons/data-viewer/blob/main/.env.example), for reference.

File: README.md
Comment (jasonaowen) [inline README.md:138]: This is no longer needed as we're not including authorization at this time, right?

```suggestion
```

File: README.md
Comment (jasonaowen) [inline README.md:140]: I'm a bit uncomfortable with this level of detail - instructions on how to configure Keycloak feels out of scope for this repo, and a lot of the steps are for authorization, which we're not doing yet. I very much appreciate the urge to document, and am happy to hear pushback, but I would suggest cutting this entire example section.

File: src/routers/opportunitiesRouter.ts
Comment (jasonaowen) [inline src/routers/opportunitiesRouter.ts:9]: Some of the routers have a `thingRouter.use(verifyJwt)` and some have `thingRouter.method(path, verifyJwt, ...)`. Why the split? Can we just use `use` everywhere?

File: src/middleware/verifyJwt.ts
Comment (jasonaowen) [inline src/middleware/verifyJwt.ts:16]: This function looks very redundant. Why `catch` an error only to immediately `throw` it? In fact, why wrap the `expressJwtMiddleware` at all? Perhaps this is an artifact of the development process that led here.

Can we simplify this to:
```suggestion
export const verifyJwt = expressjwt(jwtOptions);
```

**Review by jasonaowen — COMMENTED** [review]
I haven't found the test problem that you've been working on yet, I'm afraid!

**Comment by bickelj** [comment]
> I haven't found the test problem that you've been working on yet, I'm afraid!

That is odd but turned out to help get this resolved.

Jason and I spoke. We found that the commit above seems to work when running the service (e.g. `npm run start`) but had trouble when running the mock JWKS service at test time for many tests (e.g. `npm run test:integration`). The failures happened reliably but the errors themselves were misleading. When running with a single mock JWKS service at test time for a single test file, the failures went away. The above commit helped us diagnose the issue.

The next step is to make authentication-only branch with two atomic commits. The first commit renames awkwardly named variables. The second commit removes ad-hoc API key authentcation and replaces it with JWT authentication.

File: src/test/mockJwt.ts
Comment (bickelj) [inline src/test/mockJwt.ts:13]: More re-use of the mock JWKS server and mock JWT resolves the issue at test time. Given the re-use, it makes sense to increase the expiration duration.

File: package-lock.json
Comment (jasonaowen) [inline package-lock.json:1]: I believe the package lock file is out of date; running `npm install` on this branch leads to changes. Can you please include those changes in this commit?

File: src/openapi.json
Comment (jasonaowen) [inline src/openapi.json:11]: I tried for a while to get this to work with `"type": "openIdConnect",` as the [docs say it should](https://swagger.io/docs/specification/authentication/openid-connect-discovery/), but I wasn't able to before I reached the end of the timebox I set for myself. :disappointed:

**Review by jasonaowen — CHANGES_REQUESTED** [review]
Great work, @bickelj! I was able to run this, authenticate into the OpenAPI docs, and make authenticated requests.

We need to fix the `package-lock.json` problem, and I would appreciate enabling PKCE, but please consider everything else optional!

File: package-lock.json
Comment (bickelj) [reply package-lock.json:1]: It will pollute our production (dist) dependencies, but I will do it. Sigh.

Comment (jasonaowen) [reply package-lock.json:1]: Ah, I hadn't realize that was an intentional decision. I see now how and why you arrived at it, but unfortunately staying in that out-of-sync state is not sustainable; either developers will have to be constantly careful about not committing `package-lock.json` changes, or they'll sneak in to an unrelated commit. That sucks, and I appreciate you opening the upstream issue, but for now I think we'll just have to live with it - thank you for doing so!

File: src/auth/jwtOptions.ts
Comment (bickelj) [reply src/auth/jwtOptions.ts:26]: Instead, I will remove these lines.

File: src/openapi.json
Comment (bickelj) [reply src/openapi.json:11]: I was able to use `"type": "openIdConnect"` locally. The user experience was not good because an exhaustive litany of capabilities or options showed up. With the `oauth2` type, I was able to present exactly one known-good option.

File: src/auth/jwtOptions.ts
Comment (jasonaowen) [reply src/auth/jwtOptions.ts:26]: Thanks!

Looks like the defaults are to enable the cache and disable the rate limit ([source](https://github.com/auth0/node-jwks-rsa/blob/59373c5cd5b357fbb56b621c943f002e9f3076ca/src/JwksClient.js#L7-L14)). I think we want the rate limit on, too, to prevent a DDoS attack from spreading from the service to the IdP - right?

File: src/openapi.json
Comment (jasonaowen) [reply src/openapi.json:11]: Fair enough! Thanks for following up!

File: src/routers/documentationRouter.ts
Comment (jasonaowen) [inline src/routers/documentationRouter.ts:43]: Now that I look more closely, I think I do actually agree with Codecov here.

Adding an error handler for improperly formatted JSON is a great idea, and I'm +1 to stopping the service over it, but it's not integral to setting up auth. Let's split it out to its own commit, put the `throw` logic into the `isJsonObject` function, rename it to something like `validateOpenApiConfiguration`, and add to the existing tests!

Then, later, let's add - and test - a function that will take that already-validated configuration and substitute in the issuer. Something like:

```ts
const getSwaggerDocsWithIssuer = (documentation: JsonObject, issuer: string): JsonObject => {
  const authCodeWithNewIssuer...
  return documentationWithNewIssuer;
};
documentationRouter.use('/', swaggerUi.serve);
documentationRouter.get('/', swaggerUi.setup(getSwaggerDocsWithIssuer(documentation, issuer), options));
```

That should be a bit easier to read, much easier to test, and will address the Codecov complaint this change introduced.

**Review by jasonaowen — COMMENTED** [review]
One of the failing code coverage checks are, I believe, not worth fixing. There are two: some new code to handle the case that the OpenAPI JSON file is well-formed, and the non-promise-returning wrapper around `expressjwt`. The former is fixable - see below - but the latter seems like it would take too much effort for too little payoff.

I propose that, once this is otherwise ready, we merge anyway. @slifty, if you disagree with that plan, please speak up!

File: package-lock.json
Comment (bickelj) [reply package-lock.json:1]: I think this was really a misunderstanding on my part, though, because at the moment it doesn't appear to make any difference in `dist` or `node_modules` or the resulting docker image. I think we make no distinction and package everything from `node_modules` and `dist` at the moment. So it is more trouble than it's worth to try to pick that nit. Furthermore, I see the issue it creates. We ought to be able to `npm i` anytime without worry!

Comment (jasonaowen) [reply package-lock.json:1]: > I think we make no distinction and package everything from node_modules and dist at the moment.

That sounds like a deployment problem, but yes I agree it's out of scope for this PR.

File: src/auth/jwtOptions.ts
Comment (bickelj) [reply src/auth/jwtOptions.ts]: I recall hitting rate limits when I had `mock-jwks` set up and torn down multiple times during tests (if I recall correctly, even with only one `mock-jwks` per test suite/file). Now that we use one `mock-jwks` and one mock bearer JWT, I doubt that issue would occur with rate limits set. I will set the rate limit low during tests to see if it is exceeded (I do not expect it to, even at `1`) to double-check it is working the way I expect, and then set it to high if that works fine.

File: src/routers/documentationRouter.ts
Comment (bickelj) [reply src/routers/documentationRouter.ts:48]: Is it OK to have that commit in a separate PR? For now I can simply remove it.

File: src/auth/jwtOptions.ts
Comment (bickelj) [reply src/auth/jwtOptions.ts]: Ten per minute seemed low to me. So I set it to ten times that.

File: src/routers/documentationRouter.ts
Comment (bickelj) [reply src/routers/documentationRouter.ts]: I removed the `else` in the most recent push.

**Review by jasonaowen — APPROVED** [review]
Great work, @bickelj! Thanks for getting this to done; I know how much effort was involved.

I was able to run this locally, authenticate into the docs (using PKCE!), and make an authenticated request. I was also able to get the expected errors with an invalid token and with an expired token.

The linter is happy and CI passed. Not sure what's up with Codecov, but I'm fine to merge either without it or over its objections, and previously discussed. One small suggestion below; feel free to either fix & merge or ignore & merge.

Yay auth! :closed_lock_with_key: :balloon:

File: src/auth/jwtOptions.ts
Comment (bickelj) [reply src/auth/jwtOptions.ts:27]: Right. This is a compromise between "do not enable rate limiting" and "enable rate limiting with default value of 10 requests per minute." I think the auth server should have no trouble handling 1-2 requests per second from the service on top of the rest of the requests from data-viewer, etc. I see your point though, realistically 10 is fine too.

**Pull request merged**
